FirebaseAuth/Sources/Auth/FIRAuth.m

```diff
@@ -485,85 +491,124 @@ - (nullable instancetype)initWithAPIKey:(NSString *)APIKey
     if (![GULAppEnvironmentUtil isAppExtension]) {
       Class cls = NSClassFromString(@"UIApplication");
-      if (cls && [cls respondsToSelector:NSSelectorFromString(@"sharedApplication")]) {
+      if (cls && [cls respondsToSelector:@selector(sharedApplication)]) {
         applicationClass = cls;
       }
     }
     UIApplication *application = [applicationClass sharedApplication];
-
-    [GULAppDelegateSwizzler proxyOriginalDelegateIncludingAPNSMethods];
-    [GULSceneDelegateSwizzler proxyOriginalSceneDelegate];
-#endif  // TARGET_OS_IOS
-
-    // Continue with the rest of initialization in the work thread.
-    __weak FIRAuth *weakSelf = self;
-    dispatch_async(FIRAuthGlobalWorkQueue(), ^{
-      // Load current user from Keychain.
-      FIRAuth *strongSelf = weakSelf;
-      if (!strongSelf) {
-        return;
-      }
-      NSString *keychainServiceName =
-          [FIRAuth keychainServiceNameForAppName:strongSelf->_firebaseAppName];
-      if (keychainServiceName) {
-        strongSelf->_keychainServices =
-            [[FIRAuthKeychainServices alloc] initWithService:keychainServiceName];
-        strongSelf.storedUserManager =
-            [[FIRAuthStoredUserManager alloc] initWithServiceName:keychainServiceName];
+    if ([application respondsToSelector:@selector(isProtectedDataAvailable)]) {
+      if ([application isProtectedDataAvailable]) {
+        [self protectedDataInitialization];
+      } else {
+        // Add listener for UIApplicationProtectedDataDidBecomeAvailable.
+        self->_protectedDataDidBecomeAvailableObserver = [[NSNotificationCenter defaultCenter]
+            addObserverForName:UIApplicationProtectedDataDidBecomeAvailable
+                        object:nil
+                         queue:nil
+                    usingBlock:^(NSNotification *notification) {
+                      [self protectedDataInitialization];
+                    }];
       }
+    } else {
+      [self protectedDataInitialization];
+    }
+#else
+    [self protectedDataInitialization];
+#endif  // TARGET_OS_IOS || TARGET_OS_TV || TARGET_OS_MACCATALYST
+  }
+  return self;
+}
 
-      NSError *error;
-      NSString *storedUserAccessGroup =
-          [strongSelf.storedUserManager getStoredUserAccessGroupWithError:&error];
-      if (!error) {
-        if (!storedUserAccessGroup) {
-          FIRUser *user;
-          if ([strongSelf getUser:&user error:&error]) {
-            strongSelf.tenantID = user.tenantID;
-            [strongSelf updateCurrentUser:user byForce:NO savingToDisk:NO error:&error];
-            self->_lastNotifiedUserToken = user.rawAccessToken;
-          } else {
-            FIRLogError(kFIRLoggerAuth, @"I-AUT000001",
-                        @"Error loading saved user when starting up: %@", error);
-          }
+- (void)protectedDataInitialization {
+#if TARGET_OS_IOS || TARGET_OS_TV || TARGET_OS_MACCATALYST
+  [[NSNotificationCenter defaultCenter] removeObserver:_protectedDataDidBecomeAvailableObserver
```

Comment:
@ryanwilson Is it fine to remove an observer that was never registered (since this observer only gets registered if protected data is unavailable)? I read that it is essentially a no-op, but couldn't find a definitive answer.

Reply:
Great question! I imagine if it was never registered it would be `nil`, so you could gate it with a `nil` check. Although sounds like removing observer isn't overly necessary, but still good practice.

Reply:
Done -- tested passing `nil` directly and it was fine!